vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client/conn.go

```diff
@@ -113,13 +114,26 @@ func (c *conn) SetWriteDeadline(t time.Time) error {
 // proxy service to notify remote to drop the connection.
 func (c *conn) Close() error {
 	klog.V(4).Infoln("closing connection")
-	req := &client.Packet{
-		Type: client.PacketType_CLOSE_REQ,
-		Payload: &client.Packet_CloseRequest{
-			CloseRequest: &client.CloseRequest{
-				ConnectID: c.connID,
+	var req *client.Packet
+	if c.connID != 0 {
+		req = &client.Packet{
+			Type: client.PacketType_CLOSE_REQ,
+			Payload: &client.Packet_CloseRequest{
+				CloseRequest: &client.CloseRequest{
+					ConnectID: c.connID,
+				},
 			},
-		},
+		}
+	} else {
+		// Never received a DIAL response so no connection ID.
+		req = &client.Packet{
```

Comment:
Sorry I didn't stay on top of the original PR and sorry for jumping in here. Is this race possible: The proxy server sends the dial response to the client, and removes the dial from [pending dial](https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/da800daed6eb7fa4b505c2ac5a67136da6b4c7d8/pkg/server/server.go#L700). But before the client receives the dial response and getting the connID, conn.Close() is called and PacketType_DIAL_CLS is sent to the proxy server instead of a PacketType_CLOSE_REQ. The proxy server can't find the matching dial anymore since it's already remove from its pending dial list.

Reply:
Sure the race is possible.
KAS(Client) -> Konn Server <- Conn Agent <- Destination. If after 10 seconds the konn server gets a dial close from the KAS and very shortly there after a dial response from the agent/destination, then we hit the race and I believe still do the right thing. The dial response is dead. The KAS has stopped listening and told us its stopped listening. So the only thing we can do with that dial response at that point is make sure we clean up. If we've gotten a close dial, we can no longer complete the dial because the client has already gone away.

Reply:
We chatted offline. The behavior in this PR is a net improvement.

Before we introduced the DIAL_CLS, if the proxy client sends a CLOSE_REQ to the proxy server before receiving a DIAL_RSP, and the proxy server receives the DIAL_RSP from the proxy agent shortly after handling the CLOSE_REQ from the proxy client, then we leak:
1. an entry in the pendingDial list
2. the TCP connections from the proxy agent to the remote endpoint.

With the DIAL_CLS, we removed the leak it the pendingDial list, but we can still have the second leak.

Walter is working on adding tests to validate the existence of the second type of leak.

If the leak exists, we might be able to fix it by letting the proxy server send a CLOSE_REQ to the proxy agent if the proxy server doesn't find a pending dial matching the DIAL_RSP [here](https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/da800daed6eb7fa4b505c2ac5a67136da6b4c7d8/pkg/server/server.go#L692).